File: backend/node/api/src/application/use-cases/pause-run.ts
import type { Queue } from "../../application/queue/queue";
import { inject } from "../../infra/DI/container";

type Input = {
	runId: string;
};

export class PauseRunUseCase {
	@inject("queue")
	private declare readonly queue: Queue;

	async execute(input: Input): Promise<[true, undefined] | [undefined, Error]> {
		await this.queue.publish(
			"api.randomize",
			{ runId: input.runId },
			{ routingKey: "run.pause" },
		);

		return [true, undefined];
	}
}

File: backend/go/api/internal/run/use-cases/pause-run.go
package usecases

import (
	"context"

	"github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/queue"
	runPkg "github.com/maurodesouza/calculation-pipeline/backend/go/api/internal/run"
)

type PauseRunUseCaseDependencies struct {
	RunRepository *runPkg.RunRepository
	Queue         *queue.RabbitMQ
}

type PauseRunUseCase struct {
	PauseRunUseCaseDependencies
}

func NewPauseRunUseCase(deps PauseRunUseCaseDependencies) PauseRunUseCase {
	return PauseRunUseCase{
		PauseRunUseCaseDependencies: deps,
	}
}

type PauseRunInput struct {
	RunId string
}

func (u *PauseRunUseCase) Execute(input PauseRunInput) error {
	run, err := u.RunRepository.GetById(context.Background(), input.RunId)
	if err != nil {
		return err
	}

	if run == nil {
		return runPkg.RunNotFoundError
	}

	err = run.Pause()
	if err != nil {
		return err
	}

	err = u.RunRepository.Update(context.Background(), run)
	if err != nil {
		return err
	}

	err = u.Queue.Publish("run.pause-requested", runPkg.RunPauseRequestedMapper(run), queue.PublishConfig{})
	if err != nil {
		return err
	}

	return nil
}
